numpy/core/tests/test_einsum.py

```diff
@@ -1,284 +1,301 @@
 from __future__ import division, absolute_import, print_function
 
+import warnings
+
 import numpy as np
 from numpy.testing import (
     TestCase, run_module_suite, assert_, assert_equal, assert_array_equal,
-    assert_raises, suppress_warnings
+    assert_almost_equal, assert_raises, suppress_warnings
     )
 
+# Setup for optimize einsum
+chars = 'abcdefghij'
+sizes = np.array([2, 3, 4, 5, 4, 3, 2, 6, 5, 4, 3]) 
+global_size_dict = {}
+for size, char in zip(sizes, chars):
+    global_size_dict[char] = size
 
 class TestEinSum(TestCase):
     def test_einsum_errors(self):
-        # Need enough arguments
-        assert_raises(ValueError, np.einsum)
-        assert_raises(ValueError, np.einsum, "")
+        for do_opt in [True, False]:
+            # Need enough arguments
+            assert_raises(ValueError, np.einsum, optimize=do_opt)
+            assert_raises(ValueError, np.einsum, "", optimize=do_opt)
 
-        # subscripts must be a string
-        assert_raises(TypeError, np.einsum, 0, 0)
+            # subscripts must be a string
+            assert_raises(TypeError, np.einsum, 0, 0, optimize=do_opt)
 
-        # out parameter must be an array
-        assert_raises(TypeError, np.einsum, "", 0, out='test')
+            # out parameter must be an array
+            assert_raises(TypeError, np.einsum, "", 0, out='test', optimize=do_opt)
 
-        # order parameter must be a valid order
-        assert_raises(TypeError, np.einsum, "", 0, order='W')
+            # order parameter must be a valid order
+            assert_raises(TypeError, np.einsum, "", 0, order='W', optimize=do_opt)
 
-        # casting parameter must be a valid casting
-        assert_raises(ValueError, np.einsum, "", 0, casting='blah')
+            # casting parameter must be a valid casting
+            assert_raises(ValueError, np.einsum, "", 0, casting='blah', optimize=do_opt)
 
-        # dtype parameter must be a valid dtype
-        assert_raises(TypeError, np.einsum, "", 0, dtype='bad_data_type')
+            # dtype parameter must be a valid dtype
+            assert_raises(TypeError, np.einsum, "", 0, dtype='bad_data_type', optimize=do_opt)
 
-        # other keyword arguments are rejected
-        assert_raises(TypeError, np.einsum, "", 0, bad_arg=0)
+            # other keyword arguments are rejected
+            # Running this is python will raise this error
+            #assert_raises(TypeError, np.einsum, "", 0, bad_arg=0, optimize=do_opt)
```

Comment:
I am only getting the type error when optimize is True, not when it is False. The test seems to check both.